api/python/t4/search_util.py

```
@@ -0,0 +1,86 @@
+    )
+
+    return es_client
+
+def search(query, search_endpoint, aws_region='us-east-1'):
+    """
+    Searches your bucket. query can contain plaintext, and can also contain clauses
+    like $key:"$value" that search for exact matches on specific keys.
+
+    Returns either the request object (in case of an error)
+            or a list of objects with the following keys:
+        key: key of the object
+        version_id: version_id of object version
+        operation: Create or Delete
+        meta: metadata attached to object
+        size: size of object in bytes
+        text: indexed text of object
+        source: source document for object (what is actually stored in ElasticSeach)
+        time: timestamp for operation
+    """
+    es_client = _create_es(search_endpoint, aws_region)
+
+    payload = {'query': {'query_string': {
+        'default_field': 'content',
+        'query': query,
+        'quote_analyzer': 'keyword',
+        }}}
+
+    raw_response = es_client.search(index=ES_INDEX, body=payload)
+
+    try:
+        results = []
+        for result in raw_response['hits']['hits']:
+            key = result['_source']['key']
+            vid = result['_source']['version_id']
+            operation = result['_source']['type']
+            meta = json.dumps(result['_source']['user_meta'])
+            size = str(result['_source']['size'])
+            text = result['_source']['text']
+
+            time = str(result['_source']['updated'])
+            results.append({
+                'key': key,
+                'version_id': vid,
+                'operation': operation,
+                'meta': meta,
+                'size': size,
+                'text': text,
+                'source': result['_source'],
+                'time': time
+            })
+        results = list(sorted(results, key=lambda x: x['time'], reverse=True))
+        return results
+    except KeyError:
+        return raw_response
```

Why would returning a raw response object in this case be useful to the end user? If my function fails I'd generally expect to get an error back, not a blob.

---

no good reason; it's more helpful to the user to see whatever Elasticsearch says their problem was than to see a generic error from us

---

I doubt most users will even know what an Elasticsearch is, let alone that that's what we're using, much better here to be consistent and up-front about the error.

---

What verbiage would you recommend?

---

Seems like it should raise an exception that contains an error message with the raw response inside it?

---

I'd throw `QuiltError("Query failed unexpectedly due to either a bad query or a misconfigured search service.")` (or similar).

---

@stevededalus I've never patched an `Exception` to provide a failing response as a param (is that what you are suggesting?), though in retrospect [obviously](https://stackoverflow.com/questions/16466406/passing-an-object-with-an-exception) you can do that. Is this something that one would reasonably expect from an API?

---

No, but it's useful in `alpha` for us to debug customer issues.